hack/heap/hh_shared.c

```diff
@@ -264,31 +290,180 @@ struct timeval log_duration(const char *prefix, struct timeval start_t) {
   return end_t;
 }
 
+#endif
+
+/**************************************************************************
+
+     We create an anonymous memory file, whose `handle` might be
+     inherited by slave processes.
+
+     This memory file is tagged "reserved" but not "committed". This
+     means that the memory space will be reserved in the virtual
+     memory table but the pages will not be bound to any physical
+     memory yet. Further calls to 'VirtualAlloc' will "commit" pages,
+     meaning they will be bound to physical memory.
+
+     This is behavior that should reflect the 'MAP_NORESERVE' flag of
+     'mmap' on Unix. But, on Unix, the "commit" is implicit.
+
+     Committing the whole shared heap at once would require the same
+     amount of free space in memory (or in swap file).
+
+**************************************************************************/
+
+#ifdef _WIN32
+
+static HANDLE memfd;
+
+void memfd_init(char *shm_dir, size_t shared_mem_size) {
+  memfd = CreateFileMapping(
+    INVALID_HANDLE_VALUE,
+    NULL,
+    PAGE_READWRITE | SEC_RESERVE,
+    shared_mem_size >> 32, shared_mem_size & ((1ll << 32) - 1),
+    NULL);
+  if (memfd == NULL) {
+    win32_maperr(GetLastError());
+    uerror("CreateFileMapping", Nothing);
+  }
+  if (!SetHandleInformation(memfd, HANDLE_FLAG_INHERIT, HANDLE_FLAG_INHERIT)) {
+    win32_maperr(GetLastError());
+    uerror("SetHandleInformation", Nothing);
+  }
+}
+
+#else
+
+static int memfd = -1;
+
+void memfd_init(char *shm_dir, size_t shared_mem_size) {
+#if defined(HAVE_MEMFD_CREATE)
+  memfd = memfd_create("fb_heap", 0);
+#endif
+  if (memfd < 0) {
+#if defined(__APPLE__)
+    char memname[255];
+    sprintf(memname, "/fb_heap.%d", getpid());
+    memfd = shm_open(memname, O_CREAT | O_RDWR, 0666);
+    if (memfd < 0) {
+        uerror("shm_open", Nothing);
+    }
+#elif defined(O_TMPFILE)
+    memfd = open(shm_dir, O_TMPFILE|O_RDWR|O_CLOEXEC|O_EXCL, 0666);
```

Comment:
> > Would also like a comment about O_TMPFILE versus shm_open

Ok.

> > Also, if O_TMPFILE is not defined and we're not building on an Apple, it looks like this function will never succeed. Why not use shm_open on Linux as well?

There is always the fallback case (on Linux and Apple): `mkstemp`. It should always succeed.

I prefered `O_TMPFILE/mkstemp` over `shm_open` because it allows to move the `shared heap` from `/dev/shm` to another "disk", when `/dev/shm` is too small.